pkg/store/prometheus.go

```diff
@@ -158,6 +162,17 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
 		r.MinTime = availableMinTime
 	}
 
+	timeranges := []timerange{}
+	// Range must be at least four times smaller than the step otherwise
+	// this optimization does not make sense. Deduced this from BenchmarkQueryRangeInterval.
+	if r.Range != 0 && r.Range*4 < r.Step {
```

Comment:
Why it should be 4 times smaller? Could you explain it a little bit more?

Reply:
My logic is the following: I ran the benchmark with different conditionals:
* `r.Range < r.Step`
* `r.Range*2 < r.Step`
* `r.Range*3 < r.Step`
* `r.Range*4 < r.Step`
...

Only after `r.Range*4 < r.Step` the benchmarks showed an improvement. This is presumably because Prometheus has to perform extra work for each query i.e. locking/unlocking, hence negating any possible improvement. (https://github.com/prometheus/prometheus/blob/f678ae2146a6a54222ee16c382d6a8354473491a/storage/remote/read_handler.go#L191-L245)

Actually, I see now that Prometheus itself supports passing down Range/Step via the queries :thinking: let me try and see whether it still shows an improvement with those values simply passed down to Prometheus

Reply:
Yeah, when you simply pass down hints then there are no improvements:

```diff
diff --git a/pkg/store/prometheus.go b/pkg/store/prometheus.go
index 64e276f1..7b9a97b1 100644
--- a/pkg/store/prometheus.go
+++ b/pkg/store/prometheus.go
@@ -162,16 +162,7 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
                r.MinTime = availableMinTime
        }
 
-       timeranges := []timerange{}
-       // Range must be at least four times smaller than the step otherwise
-       // this optimization does not make sense. Deduced this from BenchmarkQueryRangeInterval.
-       if r.Range != 0 && r.Range*4 < r.Step {
-               for ts := r.MinTime; ts <= r.MaxTime; ts += r.Step {
-                       timeranges = append(timeranges, timerange{ts, ts + r.Range})
-               }
-       } else {
-               timeranges = []timerange{{r.MinTime, r.MaxTime}}
-       }
+       timeranges := []timerange{{r.MinTime, r.MaxTime}}
 
        if r.SkipChunks {
                labelMaps, err := p.client.SeriesInGRPC(s.Context(), p.base, matchers, r.MinTime, r.MaxTime)
@@ -196,7 +187,15 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
 
        queries := []*prompb.Query{}
        for _, tr := range timeranges {
-               q := &prompb.Query{StartTimestampMs: tr.mintime, EndTimestampMs: tr.maxtime}
+               q := &prompb.Query{StartTimestampMs: tr.mintime,
+                       EndTimestampMs: tr.maxtime,
+                       Hints: &prompb.ReadHints{
+                               StartMs: tr.mintime,
+                               EndMs:   tr.maxtime,
+                               StepMs:  r.Step,
+                               RangeMs: r.Range,
+                       },
+               }
                for _, m := range matchers {
                        pm := &prompb.LabelMatcher{Name: m.Name, Value: m.Value}
```

Result:
```
name                                                                  time/op
QueryRangeInterval/unoptimized_sidecar,_worst_case,_one_big_block-16  8.08ms ± 8%
QueryRangeInterval/unoptimized_sidecar,_small_blocks_with_2h_step-16  41.4ms ± 4%
QueryRangeInterval/optimized_sidecar,_worst_case,_one_big_block-16    8.01ms ± 9%
QueryRangeInterval/optimized_sidecar,_small_blocks_with_2h_step-16    39.7ms ± 3%

name                                                                  alloc/op
QueryRangeInterval/unoptimized_sidecar,_worst_case,_one_big_block-16  41.3kB ± 0%
QueryRangeInterval/unoptimized_sidecar,_small_blocks_with_2h_step-16  34.4kB ± 1%
QueryRangeInterval/optimized_sidecar,_worst_case,_one_big_block-16    41.3kB ± 1%
QueryRangeInterval/optimized_sidecar,_small_blocks_with_2h_step-16    34.5kB ± 0%

name                                                                  allocs/op
QueryRangeInterval/unoptimized_sidecar,_worst_case,_one_big_block-16     389 ± 0%
QueryRangeInterval/unoptimized_sidecar,_small_blocks_with_2h_step-16     265 ± 0%
QueryRangeInterval/optimized_sidecar,_worst_case,_one_big_block-16       389 ± 0%
QueryRangeInterval/optimized_sidecar,_small_blocks_with_2h_step-16       265 ± 0%
```